### ISSUE TITLE
Doubts about the presentation of an error

### ISSUE THREAD

**Issue by danse** [commercialhaskell/stack#4370]
A few times in the past i met the following error while using Stack:

    installed package <a> is broken due to missing package <b>

The last time the specific error was:

    installed package haddock-library-1.5.0.1 is broken due to missing package haddock-library-1.5.0.1-919WkNWGnku6Kvc4vlhbv1-attoparsec

Which has already been described in #4071. Here i am asking about the generic error, because i feel like it refers to something that i didn't learn reading the Stack documentation. Maybe it's obvious, or maybe if other people turn to have similar doubts we could think about expanding the doc or rephrasing the error.

So, package `a` is broken because package `b` is missing. Missing from where? From the resolver? Is there anything i can do, or should i rather change the resolver? The broken packages must be rebuilt ... after i changed the resolver? If i want to understand the error better do i want to look on Stack, Cabal or GHC? Or do we want people to open issues on the tracker for package `a` when this error happens?

**Comment by dbaynard** [comment]
@danse What commands did you use, which resulted in that error? It would be helpful to see the detail asked for in the issue template.

**Comment by danse** [comment]
thanks for following up @dbaynard. The error occurred in multiple cases when trying to build different projects. Originally i wanted to track up the point in the code where the error is formed, in order to understand the semantics it is referring to. If you want we can investigate together :)

**Comment by dbaynard** [comment]
Well, let's focus on a reproduction. Which stack version were you using? Then, can you reproduce it once (while running with `--verbose`) and then paste the logs?

**Comment by danse** [comment]
Oh, i fixed this problem several times, i'm not blocked on an instance of it at the moment. Often when i have been blocked on this, it turned out to be some fundamental error or some error that could be fixed with the help of package maintainers. So this is not a current issue to me, i just opened it with the speculative intent of improving the presentation of similar errors. If you prefer to consider only current issues and are collecting design conversations elsewhere, we can close this one

**Comment by dbaynard** [comment]
Let's keep this going, here, then. The issue you linked concerned stack's implementation of internal libraries, which has now been fixed, hence some of my confusion. I'm all for improving how stack handles this situation but I don't quite understand what that situation is.

> So, package `a` is broken because package `b` is missing. Missing from where? From the resolver? Is there anything i can do, or should i rather change the resolver? The broken packages must be rebuilt ... after i changed the resolver? If i want to understand the error better do i want to look on Stack, Cabal or GHC? Or do we want people to open issues on the tracker for package a when this error happens?

In this situation, `stack` should build package `b`. Unfortunately without a concrete reproduction, I can't work out the answers to any of your questions…

**Comment by danse** [comment]
So understanding the problem and improving the phrasing at a generic level seems more demanding than i had expected. The lack of a reference to a failing test case makes it harder to spot the cause and i cannot find the error text message with a text search in the code. I think that this issue is not very helpful at the moment, it can be closed. If anyone will stumble upon the problem in the future, they could find this issue and reopen the conversation with a concrete case at hand. Thanks for your support! :v:

**Comment by dbaynard** [comment]
> i cannot find the error text message with a text search in the code.

Ah, I should have said this, earlier. I believe the error comes from Cabal, but I haven't checked the source either.

OK, let's close for now, and if it reappears we can reopen. Thanks for taking the time to report this, though. And if you'd like to help make contributions then I may be able to find something you'll enjoy working on!

**Comment by danse** [comment]
> Ah, I should have said this, earlier. I believe the error comes from Cabal

oh, your comment led me to [the line where the error is defined](https://github.com/haskell/cabal/blob/b96c601d084e89361ff867a90424d4aef7fa0207/Cabal/Distribution/Backpack/Configure.hs#L191), which is informative. It seems like this message is shown to users when `Graph.broken fullIndex` is true, and that is a great clue in order to think of a clearer phrasing. So you are right, this is about Cabal and in order to progress we want to create a new issue there. This will stay here for future reference.

> I may be able to find something you'll enjoy working on

thanks for the offer! contributing to stack feels a bit daunting, and challenging this feeling would be fun! i prefer to start from ideas that pop up while using the tool though, so i guess that i will be back to this issue tracker soon! :)

**Comment by dbaynard** [comment]
Thanks for pointing out the specific source of the error in Cabal. Cabal is not responsible for dependencies, though, so I would hesitate to suggest a change there (but if you can think of an improvement, go right ahead).

Stack has volunteers who triage issues (aiming to get to them sooner than we did with you — thank you for your patience!). We're looking for more people to get involved, so if that appeals let one of us know. There are also issues marked 'help wanted' and 'newcomer friendly'. Whatever you decide, I look forward to it!

**Comment by danse** [comment]
> Stack has volunteers who triage issues (aiming to get to them sooner than we did with you — thank you for your patience!). We're looking for more people to get involved, so if that appeals let one of us know. There are also issues marked 'help wanted' and 'newcomer friendly'. Whatever you decide, I look forward to it!

Thanks, i'll keep that in mind! :)
